### PR TITLE
Add json output for diff command

### LINKED ISSUE #2508: json output for diff
Output of `restic version`
--------------------------
0.9.6

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
diff output recognize the --json flag

What are you trying to do?
--------------------------
when trying to scale restic we run --quiet on backups with millions of files on NFS mounts to eliminate the scan.  however that mean we do not have results of the backup for later processing/monitoring.  So diff provides kind of a workaround to get at least some of the statistics.  It would be nice if that was json though and possibly also an option not to print the 1000's of +, -, M lines just the file/dir etc changes.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
love it!

### PR DESCRIPTION
**Pull request by vgt**
What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR introduces support for the --json flag for the diff command.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #2508 

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.
- [x] Squash commits(?)

### CODE REVIEW

**Comment by vgt** [comment]
I am not sure about test cases for the change. Since this is the first time I use go, and there are no preexisting test cases for the diff command, I am not able to cheat myself out of this.
For the documentation in the manual, I checked how the --json flag is documented in other commands and it is only referenced in the general "--json outputs json for all supported commands".
Can someone give me some feedback on what needs to be added for this PR to be even considered?

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Are you aware of #2542? As explained in https://github.com/restic/restic/pull/2542#issuecomment-877796807 I don't think the output format is suitable for snapshots with lots of files in them. Please see the comment in https://github.com/restic/restic/issues/2508#issuecomment-1001683455 for my output format suggestion.

File: cmd/restic/cmd_diff.go [hunk 1]
```diff
@@ -115,19 +116,24 @@ func addBlobs(bs restic.BlobSet, node *restic.Node) {
...
-		BlobsCommon: restic.NewBlobSet(),
-	}
+func NewDiffStats() (*DiffStats, *BlobSets, []Change) {
+	return &DiffStats{},
+		&BlobSets{
+			BlobsBefore: restic.NewBlobSet(),
+			BlobsAfter:  restic.NewBlobSet(),
+			BlobsCommon: restic.NewBlobSet(),
+		},
+		make([]Change, 0)
```
[the remark this answers is not in the document]
Comment (vgt) [reply cmd/restic/cmd_diff.go:136]: I did not see that point, but it makes a lot of sense.

[the remark this answers is not in the document]
Comment (vgt) [reply cmd/restic/cmd_diff.go]: I have changed the implementation to be more like what I found in the ls command. Not really happy that the output is not a valid JSON, but it is at least "consistent",

File: changelog/unreleased/issue-2508 [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Enhancement #2508: Support JSON output for diff
```
[the remark this answers is not in the document]
Comment (vgt) [reply changelog/unreleased/issue-2508]: Changed that to be the same as other changes. :)

File: cmd/restic/cmd_diff.go [hunk 3]
```diff
@@ -348,7 +378,57 @@ func runDiff(opts DiffOptions, gopts GlobalOptions, args []string) error {
...
+			if err != nil {
+				Warnf("JSON encode failed: %v\n", err)
+			}
+		}
+	} else {
+		printChange = func(path string, mode string) {
+			Printf("%-5s%v\n", mode, path)
+		}
+
+		printDiffStats = func(stats *DiffStats) {
```
Comment (MichaelEischer) [inline cmd/restic/cmd_diff.go:421]: Please inline the `printDiffStats` methods and just add an if at the end of `runDiff`.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
I guess my reference regarding the output format in https://github.com/restic/restic/issues/2508#issuecomment-1001683455 was not explicit enough, sorry. Please follow that format unless there is a (good) reason not to.

Once you're finished, please also squash the commits (unless you want to split the changes into several steps).

File: cmd/restic/cmd_diff.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (MichaelEischer) [reply cmd/restic/cmd_diff.go]: It is not a single JSON document. But rather a stream of JSON documents separated by new lines. That type of streaming format is rather common. Tools like `jq` also provide explicit support for it.

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks for implementing this feature!

**Comment by vgt** [comment]
> LGTM. Thanks for implementing this feature!

Thanks for your patience guiding through my first ever open source contribution. I just fixed a minor thing lint complained about in integration_test.go. Though it seems I cannot start the workflow to run again as far as I can tell.

**Comment by MichaelEischer** [comment]
> Though it seems I cannot start the workflow to run again as far as I can tell.

Github has a restriction since several months that ci test for a merge request are not run automatically if it is the first contribution to a repository. This restriction is probably in place to fend of people trying to use the Github CI to mine some crypto coins.

**Pull request merged**